crates/hir-def/src/hir/type_ref.rs

```diff
@@ -104,35 +109,90 @@ impl TraitRef {
     }
 }
 
+thin_vec_with_header_struct! {
+    pub new(pub(crate)) struct FnType, FnTypeHeader {
+        pub params: [(Option<Name>, TypeRefId)],
+        pub is_varargs: bool,
+        pub is_unsafe: bool,
+        pub abi: Option<Symbol>; ref,
+    }
+}
+
+#[derive(Clone, PartialEq, Eq, Hash, Debug)]
+pub struct ArrayType {
+    pub ty: TypeRefId,
+    // FIXME: This should be Ast<ConstArg>
+    pub len: ConstRef,
+}
+
+#[derive(Clone, PartialEq, Eq, Hash, Debug)]
+pub struct RefType {
+    pub ty: TypeRefId,
+    pub lifetime: Option<LifetimeRef>,
+    pub mutability: Mutability,
+}
+
 /// Compare ty::Ty
-///
-/// Note: Most users of `TypeRef` that end up in the salsa database intern it using
-/// `Interned<TypeRef>` to save space. But notably, nested `TypeRef`s are not interned, since that
-/// does not seem to save any noticeable amount of memory.
 #[derive(Clone, PartialEq, Eq, Hash, Debug)]
 pub enum TypeRef {
     Never,
     Placeholder,
-    Tuple(Vec<TypeRef>),
+    Tuple(EmptyOptimizedThinVec<TypeRefId>),
     Path(Path),
-    RawPtr(Box<TypeRef>, Mutability),
-    Reference(Box<TypeRef>, Option<LifetimeRef>, Mutability),
-    // FIXME: This should be Array(Box<TypeRef>, Ast<ConstArg>),
-    Array(Box<TypeRef>, ConstRef),
-    Slice(Box<TypeRef>),
+    RawPtr(TypeRefId, Mutability),
+    Reference(Box<RefType>),
+    Array(Box<ArrayType>),
+    Slice(TypeRefId),
     /// A fn pointer. Last element of the vector is the return type.
-    Fn(
-        Box<[(Option<Name>, TypeRef)]>,
-        bool,           /*varargs*/
-        bool,           /*is_unsafe*/
-        Option<Symbol>, /* abi */
-    ),
-    ImplTrait(Vec<Interned<TypeBound>>),
-    DynTrait(Vec<Interned<TypeBound>>),
+    Fn(FnType),
+    ImplTrait(ThinVec<TypeBound>),
+    DynTrait(ThinVec<TypeBound>),
     Macro(AstId<ast::MacroCall>),
     Error,
 }
 
+#[cfg(target_arch = "x86_64")]
+const _: () = assert!(size_of::<TypeRef>() == 16);
```

Comment:
I benchmarked memory usage using jemalloc, which I think includes the allocator overhead. Speed could be affected, but I don't think it will (nor did my benchmarks showed), since boxed types account for less than 10% of all types. If we allow 24 bytes we are able to unbox `Reference` but not `Array`, but most of the 10% is references anyway. I will try to benchmark memory usage of each option to see what is better.

Reply:
For context: here are the statistics of the r-a repo:

total_types: 656222 (100%)
tuples: 11492 (1.75%)
path_types: 552122 (84.14%)
raw_ptrs: 13167 (2.01%)
references: 58745 (8.95%)
references_with_lifetimes: 10529 (1.60%)
arrays: 2792 (0.43%)
slices: 6691 (1.02%)
fns: 821 (0.13%)
impl_traits: 1523 (0.23%)
dyn_traits: 4723 (0.72%)
macros: 1050 (0.16%)

This shows that if we are really concerned by the boxing, we can split `Reference` into `ReferenceWithLifetime` and `ReferenceWithoutLifetime` and box only the latter, and then we'll have a bit more than 2% boxed.